### PR TITLE
[Fiber] Refactor bailoutOnFinishedWork

### PR DESCRIPTION
**Pull request by sebmarkbage**
See individual commits.

This consolidates bailouts on finished finished work into each branch behind the tag.

### CODE REVIEW

**Comment by sebmarkbage** [comment]
I'm still looking at bailoutOnLowPriority which also has inconsistent pushes. Want to see if we can make it avoid calling completeWork in that case.

File: src/renderers/shared/fiber/ReactFiberScheduler.js [hunk 1]
```diff
@@ -200,6 +206,14 @@ module.exports = function<T, P, I, TI, C, CX>(config : HostConfig<T, P, I, TI, C
     if (highestPriorityRoot) {
       nextPriorityLevel = highestPriorityLevel;
       priorityContext = nextPriorityLevel;
+
+      // Before we start any new work, let's make sure that we have a fresh
+      // stack to work from.
+      // TODO: This call is burried a bit too deep. It would be nice to have
+      // a single point which happens right before any new work and
+      // unfortunately this is it.
+      resetContextStack();
```
Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberScheduler.js]: Could we put the reassignment of `priorityContext` (line 208) inside `resetContextStack`? It's not quite the same as the other contexts but it serves a similar purpose.

File: src/renderers/shared/fiber/ReactFiberBeginWork.js [hunk 2]
```diff
@@ -181,26 +181,33 @@ module.exports = function<T, P, I, TI, C, CX>(
...
+    const memoizedProps = workInProgress.memoizedProps;
+    if (hasContextChanged()) {
+      // Normally we can bail out on props equality but if context has changed
+      // we don't do the bailout and we have to reuse existing props instead.
+      if (nextProps === null) {
+        nextProps = current && current.memoizedProps;
       }
+    } else if (nextProps === null || memoizedProps === nextProps || (
+        // TODO: Disable this before release, since it is not part of the public API
+        // I use this for testing to compare the relative overhead of classes.
```
Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: Will we actually do this?

File: src/renderers/shared/fiber/ReactFiberBeginWork.js [hunk 3]
```diff
@@ -274,11 +274,17 @@ module.exports = function<T, P, I, TI, C, CX>(
...
       const prevState = workInProgress.memoizedState;
       const state = beginUpdateQueue(workInProgress, updateQueue, null, prevState, null, priorityLevel);
+      if (prevState === state) {
+        // If the state is the same as before, that's a bailout because we had
+        // no work matching this priority.
+        return bailoutOnAlreadyFinishedWork(current, workInProgress);
+      }
       const element = state.element;
       reconcileChildren(current, workInProgress, element);
       workInProgress.memoizedState = state;
```
Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: Shouldn't we return `workInProgress.child` inside this branch? I get that `bailoutOnAlreadyFinishedWork` does that anyway but if there's state then it's not really a bailout, is it?

File: src/renderers/shared/fiber/ReactFiberUpdateQueue.js [hunk 4]
```diff
@@ -80,15 +80,6 @@ function comparePriority(a : PriorityLevel, b : PriorityLevel) : number {
...
 }
 
-function hasPendingUpdate(queue : UpdateQueue, priorityLevel : PriorityLevel) : boolean {
-  if (!queue.first) {
-    return false;
-  }
-  // Return true if the first pending update has greater or equal priority.
-  return comparePriority(queue.first.priorityLevel, priorityLevel) <= 0;
-}
-exports.hasPendingUpdate = hasPendingUpdate;
```
Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberUpdateQueue.js]: Nice!

File: src/renderers/shared/fiber/ReactFiberBeginWork.js [hunk 5]
```diff
@@ -251,9 +270,73 @@ module.exports = function<T, P, I, TI, C, CX>(
...
+      const prevState = workInProgress.memoizedState;
+      const state = beginUpdateQueue(workInProgress, updateQueue, null, prevState, null, priorityLevel);
+      if (prevState === state) {
+        // If the state is the same as before, that's a bailout because we had
+        // no work matching this priority.
+        return bailoutOnAlreadyFinishedWork(current, workInProgress);
+      }
+      const element = state.element;
+      reconcileChildren(current, workInProgress, element);
+      workInProgress.memoizedState = state;
```
Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: GitHub suppressed [my comment on an older commit](https://github.com/facebook/react/pull/8613#discussion_r93501718) so here it is again:

Shouldn't we return `workInProgress.child` inside this branch? I get that `bailoutOnAlreadyFinishedWork` does that anyway but if there's state then it's not really a bailout, is it?

File: src/renderers/shared/fiber/ReactFiberScheduler.js (hunk 1 above)
Comment (sebmarkbage) [inline src/renderers/shared/fiber/ReactFiberScheduler.js]: We could but I suspect that with #8611 these will be unified to a single call to an external module which is why I grouped them inside of just inlining them.

File: src/renderers/shared/fiber/ReactFiberBeginWork.js (hunk 3 above)
Comment (sebmarkbage) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: Ah yes. Good catch.

File: src/renderers/shared/fiber/ReactFiberBeginWork.js (hunk 2 above)
Comment (sebmarkbage) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: ¯\_(ツ)_/¯

File: src/renderers/shared/fiber/ReactFiberBeginWork.js (hunk 5 above)
Comment (sebmarkbage) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: Yea. Fixed.

File: src/renderers/shared/fiber/ReactFiberScheduler.js (hunk 1 above)
Comment (bvaughn) [inline src/renderers/shared/fiber/ReactFiberScheduler.js]: That is correct. At least in its current form, #8611 unifies these context stacks and it is no longer possible to reset them independently. You can only _reset all_ stacks or _unwind all_ stacks to a current `Fiber`.

**Pull request merged**